src/ProvidedTypes.fs

```diff
@@ -8037,8 +8037,68 @@ namespace ProviderImplementation.ProvidedTypes
         override x.ToString() = "tgt assembly " + x.FullName
 
 
+
+    type ProvidedAssembly(isTgt: bool, assemblyName:AssemblyName, assemblyFileName: string) =
+      
+        inherit Assembly()
+        let theTypes = ResizeArray<ProvidedTypeDefinition[] * string list option>()
+        
+        let addTypes (ptds:ProvidedTypeDefinition[], enclosingTypeNames: string list option) =
+            for pt in ptds do
+                if pt.IsErased then failwith ("The provided type "+pt.Name+"is marked as erased and cannot be converted to a generated type. Set 'IsErased=false' on the ProvidedTypeDefinition")
+                if not isTgt && pt.BelongsToTargetModel then failwithf "Expected '%O' to be a source ProvidedTypeDefinition. Please report this bug to https://github.com/fsprojects/FSharp.TypeProviders.SDK/issues" pt
+                if isTgt && not pt.BelongsToTargetModel then failwithf "Expected '%O' to be a target ProvidedTypeDefinition. Please report this bug to https://github.com/fsprojects/FSharp.TypeProviders.SDK/issues" pt
+            theTypes.Add (ptds, enclosingTypeNames)
+
+        let theTypesArray = lazy (theTypes.ToArray() |> Array.collect (function (ptds, None) -> Array.map (fun ptd -> (ptd :> Type)) ptds | _ -> [| |]))
+
+        override x.GetReferencedAssemblies() = notRequired x "GetReferencedAssemblies" (assemblyName.ToString())
+
+        override __.GetName() = assemblyName
+
+        override __.FullName = assemblyName.ToString()
+
+        override __.Location = assemblyFileName
+
+        override __.ReflectionOnly = true
+
+        override __.GetTypes () = theTypesArray.Force()
+
+        override x.GetType (nm: string) = 
+            if nm.Contains("+") then
+                let i = nm.LastIndexOf("+")
+                let enc,nm2 = nm.[0..i-1], nm.[i+1..]
+                match x.GetType(enc) with
+                | null -> null
+                | t -> t.GetNestedType(nm2,bindAll)
+            else
+                theTypesArray.Force() 
+                |> Array.tryPick (fun ty -> if ty.FullName = nm then Some ty else None) 
+                |> Option.toObj
+        
+        new () = 
+            let assemblyFileName = Path.ChangeExtension(Path.GetTempFileName(), "dll")
+            let simpleName = Path.GetFileNameWithoutExtension(assemblyFileName)
+            ProvidedAssembly(AssemblyName(simpleName), assemblyFileName)
+
+        new (assemblyName, assemblyFileName) = 

```

Comment:
Previously, when passing a filename, you didn't need to specify the assemblyname - It seems like this is almost always going to be `Path.GetFileNameWithoutExtension` off the `assemblyFileName` - should it just be taken care of for the caller, so they don't need to build the assembly name as well?  (That's closer to the old API - you just passed one string for path)

Reply:
I like the parameterless constructor, but I do not like the implementation of [Path.GetTempFileName](https://msdn.microsoft.com/en-us/library/system.io.path.gettempfilename(v=vs.110).aspx)

> Creates a uniquely named, zero-byte temporary file on disk and returns the full path of that file.

> The GetTempFileName method will raise an IOException if it is used to create more than 65535 files without deleting previous temporary files.

Once, I already wrote the code that calls `Path.GetTempFileName` too often and did not delete this empty file... This is weird.

Not sure if it is the case... But in theory, TP may crash in looong running VS/IDE instance 😄

Reply:
@sergey-tihon Yes, you're correct, we are creating empty files, that is wrong